workers/unity/Assets/Gdk/Core/GameObjectRepresentation/GameObjectReference.cs

```diff
@@ -1,9 +1,37 @@
+using Improbable.Worker;
 using UnityEngine;
 
 namespace Improbable.Gdk.Core
 {
+    /// <summary>
+    ///     Invokes dispatcher callbacks on GameObjects that represent entities.
+    /// </summary>
     public class GameObjectReference : Component
     {
         public GameObject GameObject;
+
+        public void OnAddComponent(uint componentId)
+        {
+        }
+
+        public void OnRemoveComponent(uint componentId)
+        {
+        }
+
+        public void OnAuthorityChange(uint componentId, Authority authority)
+        {
+        }
+
+        public void OnComponentUpdate(uint componentId, object update)
```

Comment:
Whatever invokes the callbacks is allowed to know about readers, so that can use the actual types.

Reply:
@samiwh The idea is for GameObjectReference to be not codegenned, though. We could work around that by exposing a method to get a list of ReaderWriters for a given component ID maybe? Then the translation classes would query GameObjectReference for that list (which would exist internally anyway), cast it to the specific type it knows them to be and call the specific type callbacks itself.

Reply:
Might be easier to have the readers and writers exposed to something that the codegen code can use. Have codegened helper methods that match against all entities that have the relevant codegened components and a handle to the readers and writers of a given type. Then directly call the callbacks on the reader/writer.

Using object isn't something we can keep.

Reply:
Ok, I have two ideas here:

1. Let the `GameObjectReference` callbacks not contain payloads, i.e. change `public void OnComponentUpdate(uint componentId, object update)` to `public void OnComponentUpdate(uint componentId)`. Readers/Writers do not get the update, they just get notified that an update happened. Readers/Writers fetch the update payload themselves using `EntityManager.GetComponentObject<specific codegened type>(entity)`.

2. Have codgenerated extension methods for the GameObjectReference class for each component. Things will get very ugly when you have multiple components with the same name but in different namespaces.

3. Generate a type specific GameObjectReference class for every component and add multiple of those to entities. The sad part here is that we still need to keep a general component unspecific GameObjectReference to handle the injection/lifecycle.

Reply:
I like the first one. I can't see any serious downside other than the fact that the code looks a bit surprising to read? Which could be mitigated by comments above the GameObjectReference methods.

Reply:
Option 2 looks like a no-go. Option 1 is my favourite right now because it's the simplest and fastest to implement. Also, we'd need to implement Option 1 as a prerequisite for implementing Option 3.